File: Medium/Allocate minimum number of pages/allocate-minimum-number-of-pages.py
class Solution:
    
    #Function to find minimum number of pages.
    def findPages(self,A, N, M):
        beg = max(A)
        end = sum(A) 
        ans =  0 
        while beg <= end:
            mid  = (beg + end) // 2
            # distribution
            c = 1
            p =  0
            for i in range(N):
                p  += A[i]
                if p  > mid:
                    c +=1
                    p = A[i]
                    # break
            if c <= M:
                # for checking the diffrences betweeen the students
                ans = mid 
                end  = mid - 1
            else:
                beg = mid + 1
        if M > N:
            return -1
        else:
            
            return ans



#{ 
 # Driver Code Starts
#Initial Template for Python 3

#contributed by RavinderSinghPB
if __name__=='__main__':
    t=int(input())
    for _ in range(t):
        
        n=int(input())
        
        arr=[int(x) for x in input().strip().split()]
        m=int(input())
        
        ob=Solution()
        
        print(ob.findPages(arr,n,m))
# } Driver Code Ends

File: Medium/Allocate minimum number of pages/allocate-minimum-number-of-pages.cpp
//{ Driver Code Starts
// Initial template for C++

#include<bits/stdc++.h>
using namespace std;


// } Driver Code Ends
//User function template in C++

class Solution 
{
    public:
    bool isPossible(int A[], int n, int m, int mid){
        int studentsReq = 1;
        int curr_sum = 0;
        
        for(int i=0; i<n; i++){
            if(A[i]>mid){
                return false;
            }
            
            if(curr_sum+A[i]<=mid){
                curr_sum+=A[i];
            }
            else{
                studentsReq++;
                if(studentsReq>m){
                    return false;
                }
                curr_sum = A[i];
            }
        }
        return true;
    }
    //Function to find minimum number of pages.
    int findPages(int A[], int n, int m) 
    {
        if(n<m){
            return -1;
        }
        int s = 0;
        int ans = -1;
        int sum = 0;
        for(int i=0; i<n; i++){
            sum+=A[i];
        }
        int e = sum;
        int mid = s + (e-s)/2;
        while(s<=e){
            if(isPossible(A,n,m,mid)){
                ans = mid;
                e = mid - 1;
            }
            else{
                s = mid + 1;
            }
            mid = s + (e-s)/2;
        }
        return ans;
    }
};

//{ Driver Code Starts.

int main() {
    int t;
    cin>>t;
    while(t--){
        int n;
        cin>>n;
        int A[n];
        for(int i=0;i<n;i++){
            cin>>A[i];
        }
        int m;
        cin>>m;
        Solution ob;
        cout << ob.findPages(A, n, m) << endl;
    }
    return 0;
}

// } Driver Code Ends